### PR TITLE
[IA-2469] preventing csrf attacks in Leo proxy

### PR DESCRIPTION
**Pull request by gpcarr**
jira ticket: https://broadworkbench.atlassian.net/browse/IA-2469

accompanying firecloud-develop PR: https://github.com/broadinstitute/firecloud-develop/pull/2408

added unit tests and tested in a fiab manually adding `refererConfig` to [reference.conf](https://github.com/DataBiosphere/leonardo/pull/1854/files#diff-9024d8d3f171d065fa3b61dbc7dc2c56bdc6587644fde470722fdc92e7068f61R445). the config that was manually added is identical to the config added to firecloud-develop in this [PR](https://github.com/broadinstitute/firecloud-develop/pull/2408) for the dev environment. Just kicked off a `fiab-start` job that will test these changes alongside the firecloud-develop changes as well

---
Have you read [CONTRIBUTING.md](https://github.com/DataBiosphere/leonardo/blob/develop/CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [ ] I've documented my API changes in Swagger

In all cases:

- [ ] Get a thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green
- [ ] Squash and merge; Delete your branch after this
- [ ] Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

**Comment by gpcarr** [comment]
failing tests include:
`should 404 for non-existent runtimes *** FAILED ***`
`should proxy websockets *** FAILED ***`
`should 404 for non-existent apps *** FAILED ***`

**Comment by gpcarr** [comment]
> failing tests include:
> `should 404 for non-existent runtimes *** FAILED ***`
> `should proxy websockets *** FAILED ***`
> `should 404 for non-existent apps *** FAILED ***`

actually only failing test is `should proxy websockets`
not sure if websockets need to be handled differently in terms of referers @rtitle do you have any insight? could just need to add header to this [line](https://github.com/DataBiosphere/leonardo/blob/develop/http/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/http/api/ProxyRoutesSpec.scala#L231) but want to make sure that there's nothing else i'm missing

File: http/src/main/resources/reference.conf
Comment (rtitle) [inline http/src/main/resources/reference.conf]: Reading the code, it looks like `validPorts` is only used if the host == `"localhost"`. Wondering if that might be confusing to remember -- could we just represent the allow-list as host:port strings? For example:
```
refererConfig {
  allowedHosts = [
    "localhost:3000",
    "app.terra.bio"
  ]
}
```

Then we could probably check the referer against it with `referer.getUri.authority.toString`:
```
scala> val uri1 = Uri("https://app.terra.bio/")
val uri1: akka.http.scaladsl.model.Uri = https://app.terra.bio/

scala> uri1.authority.toString
val res13: String = app.terra.bio

scala> val uri2 = Uri("http://localhost:3000")
val uri2: akka.http.scaladsl.model.Uri = http://localhost:3000

scala> uri2.authority.toString
val res14: String = localhost:3000
```

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/config/RefererConfig.scala
Comment (rtitle) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/config/RefererConfig.scala]: Not certain we need this ADT -- 1 case class is probably fine:
```
final case class RefererConfig(allowedHosts: List[String])
```

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/api/ProxyRoutes.scala
Comment (rtitle) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/api/ProxyRoutes.scala]: Reminder to remove logging before merge

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/api/ProxyRoutes.scala
Comment (rtitle) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/api/ProxyRoutes.scala]: Nit: could maybe be more concise:
```
optionalHeaderValueByType(Upgrade) flatMap {
  case Some(upgrade) if upgrade.hasWebSocket => pass
  case _ => checkReferer
}
```

File: http/src/main/resources/reference.conf
Comment (dinvlad) [reply http/src/main/resources/reference.conf]: Sounds great! Any approach where host:port can be determined in full (without possibility to spoof the check with something like `app.terra.bio.evil.org`) is good.

**Comment by gpcarr** [comment]
@rtitle just pushed up changes to address comments, going to test in a fiab now :)

File: http/src/main/resources/reference.conf
Comment (rtitle) [inline http/src/main/resources/reference.conf]: As mentioned in 1-1, can probably put dev or example hosts here instead of "none"

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/config/RefererConfig.scala
Comment (rtitle) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/config/RefererConfig.scala]: Nit: maybe make this a `Set[String]` since we're checking it with `.contains()`

**Comment by rtitle** [comment]
Couple more nits, overall looks good

File: http/src/main/resources/reference.conf
Comment (gpcarr) [reply http/src/main/resources/reference.conf]: will update!

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/config/RefererConfig.scala
Comment (gpcarr) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/config/RefererConfig.scala]: ah yes I also like that better - will update!

File: http/src/main/resources/reference.conf
Comment (rtitle) [inline http/src/main/resources/reference.conf]: I think this should be `leonardo-fiab.dsde-dev.broadinstitute.org:30443`? 

(Since fiabs have a port)

**Review by rtitle — APPROVED** [review]
Looks good - one fiab host comment, otherwise should be good to merge once tests pass

**Comment by rtitle** [comment]
Hm.. just thought of something. Do automation tests work? Those use Selenium to hit the proxy. Do those calls have a referrer, and can we allow that somehow?

File: http/src/main/resources/reference.conf
Comment (gpcarr) [reply http/src/main/resources/reference.conf]: going to retest in my fiab by running the hotswap script since fiab-start keeps failing

**Comment by gpcarr** [comment]
> Hm.. just thought of something. Do automation tests work? Those use Selenium to hit the proxy. Do those calls have a referrer, and can we allow that somehow?

this is a good point - i'll log some stuff to see what the referer is for automation tests. automation tests run on fiabs correct? so my guess is we should be fine but i'll double check

**Comment by gpcarr** [comment]
unrelated test failure (only 1 test failed `NotebookGCECustomizationSpec` `should execute user-specified start script` which is a currently known flaky test)

jenkins re-test

**Comment by gpcarr** [comment]
jenkins retest

**Comment by gpcarr** [comment]
same unrelated, flakey test failure from before ... jenkins retest

**Review by rtitle — APPROVED** [review]
This looks good to me.. 👍  if tests pass

**Comment by gpcarr** [comment]
> This looks good to me.. 👍 if tests pass

tests passed - need to wait for my firecloud-develop PR to be unblocked before merging this PR tho

File: http/src/main/resources/reference.conf
Comment (Qi77Qi) [inline http/src/main/resources/reference.conf]: this list can be removed too?

**Review by Qi77Qi — APPROVED** [review]

**Comment by gpcarr** [comment]
jenkins retest - same failure as before

**Comment by gpcarr** [comment]
unrelated test failure ... jenkins retest

**Pull request merged**
